### DISCUSSION TITLE
microsoft.dotnet.designtools.client.proxies.ControlProxy not available

### DISCUSSION THREAD

**Discussion by AdmiralSnyder** [dotnet/winforms#7696] [Q&A] [answered]
### Environment

VS 2022 Version 17.4.0 Preview 1.0
windows 11, 64bit

### .NET version

since it's a VS extension, it targets Framework 4.8

### Did this work in a previous version of Visual Studio and/or previous .NET release?

sure, in VS2019 it's been working (i don't remember when the switch to the new oop designer happened, but _that_ is the relevant breaking point.

### Issue description

i'm trying to migrate a VS extension to VS2022 that is using DTE to get to the currently selected control in the winforms designer, and toggles through its Dock property, so that a control can be snapped to the sides of its parent via hotkey.

in VS2022, the out of process designer is being used, and the object that is selected (that DTE / ISelectionService returns) is an
microsoft.dotnet.designtools.client.proxies.ControlProxy<T>

i cannot find this type. is it not publicly visible? are there interfaces i could cast it to?
basically, i need to get to the underlying control to set its dock property.

Any help is greatly appreciated.

### Steps to reproduce

i can put the extension project onto github and make it publicly available, if requested.

**Comment by AdmiralSnyder** [comment]
having a look at the assembly of the GetType() led me to the `Microsoft.DotNet.DesignTools.Clients.dll` assembly in the VS directory `C:\Program Files\Microsoft Visual Studio\2022\Preview\Common7\IDE\CommonExtensions\Microsoft\Windows.Forms\Microsoft.DotNet.DesignTools.Client.dll` which i can reference... BUT how do i reference that assembly _correctly_ for vsix packaging? I suspect I can't just bundle it (nor do i want to).

working with those proxy types is straightforward, though - the functionality of the extension is woking again :-)

**Comment by merriemcgaw** [comment]
We are still in the process of producing documentation around the out of process designer SDK. The latest info can be found [here](https://github.com/dotnet/winforms/discussions/7073) and further samples and instructions will be available soon. The sample that is referenced in this thread will show you how to reference the WinForms Designer Extensibility SDK, which should automatically add the reference to the DLL you are talking about.

**Comment by AdmiralSnyder** [comment]
thanks for that link. the linked winforms designer sdk nuget package was the exact thing i was looking for :-)
(shameless plug: https://marketplace.visualstudio.com/items?itemName=AlexanderGayko.WinformsDesignerHelpers )

**Comment by KlausLoeffelmann** [comment, marked as answer]
We are having this blog post as the primary source of documentation for Custom Control Designers in the Out-Of-Process WinForms Designer:

https://devblogs.microsoft.com/dotnet/custom-controls-for-winforms-out-of-process-designer/

Please note, that we're working on additional docs at the moment, but it's going to take a while. 
We're planning to release the final first version of the Designer SDK around the time Visual Studio 17.6 GA will be released.
